File: control calcs/rocket_sim.m
% Example MATLAB script to run the rocket simulation and plot altitude vs time

close all; clear; clc;

% ---------------------------
% Define constants
% ---------------------------
gravity = -9.81;             % m/s²
burnout_time = 0.5;           % Example burnout time in seconds
apogee = 1000;               % Example desired apogee in metres
terminal_velocity = -8;     % Example terminal velocity in m/s
resolution = 320 * 50;          % Number of data points

% Atmospheric and gas constants
temp_lapse_rate = -0.0065;  % K/m (typical environmental lapse rate)
temperature = 300;       % K (approx. 15°C at sea level)
h_b = 0;                    % Base altitude in metres
static_pressure = 101325;   % Pa (standard sea level pressure)
molar_mass = 0.0289644;     % kg/mol (molar mass of Earth's air)
gas_constant = 8.3144598;   % J/(mol*K) (universal gas constant)

% Run the simulation
data = runSimulation(burnout_time, apogee, terminal_velocity, gravity, ...
                     resolution, temp_lapse_rate, temperature, h_b, ...
                     static_pressure, molar_mass, gas_constant);

% Plot the altitude vs. time
figure;
plot(data.time, data.altitude, 'LineWidth', 1.5);
xlabel('Time (s)');
ylabel('Altitude (m)');
title('Altitude vs. Time');
grid on;

% ---------------------------
% Functions
% ---------------------------

function y = compute_a_b(g, t_b, S_a)
    % Translated from C++ logic to MATLAB
    term1 = -g * t_b^2;
    sqrt_argument = (term1 * term1) - (4.0 * t_b^2 * 2.0 * g * S_a)

    if sqrt_argument < 0.0
        y = NaN;
        return;
    end

    sqrt_value = sqrt(sqrt_argument);
    numerator = g * t_b^2 + sqrt_value;
    denominator = 2.0 * t_b^2;
    y = numerator / denominator;
end

function P = altitudeToPressure(h, temp_lapse_rate, temperature, h_b, gravity, molar_mass, gas_constant, static_pressure)
    base = 1 + ((temp_lapse_rate / temperature) * (h - h_b));
    exponent = (gravity * molar_mass) / (gas_constant * temp_lapse_rate);
    P = static_pressure * (base^exponent);
end

function data = runSimulation(burnout_time, apogee, terminal_velocity, gravity, resolution, ...
                              temp_lapse_rate, temperature, h_b, static_pressure, molar_mass, gas_constant)
    % Compute initial acceleration
    acceleration = compute_a_b(gravity, burnout_time, apogee)

    % Compute key times
    time_apogee = abs((burnout_time * (acceleration - gravity)) / (-gravity));
    time_a_t = abs(terminal_velocity / gravity);
    dist_t = apogee + ((terminal_velocity^2) / (2 * gravity));
    time_t_l = abs(dist_t / terminal_velocity);
    time_total = time_apogee + time_a_t + time_t_l;

    if time_total <= 0
        data = [];
        return;
    end

    % Preallocate arrays
    data.velocity = zeros(1, resolution);
    data.altitude = zeros(1, resolution);
    data.time = zeros(1, resolution);
    data.pressure = zeros(1, resolution);

    % Initial conditions
    data.velocity(1) = 0.0;
    data.altitude(1) = 0.0;
    data.time(1) = 0.0;

    prev_t = 0.0;
    acc_current = acceleration;

    % Simulation loop
    for i = 2:resolution
        % Map index to time linearly from 0 to time_total
        data.time(i) = ((i - 2) / (resolution - 2)) * time_total;
        dt = data.time(i) - prev_t;

        % After burnout, adjust acceleration and velocity if terminal velocity is reached
        if data.time(i) >= burnout_time
            if data.velocity(i - 1) <= terminal_velocity
                acc_current = 0;
                data.velocity(i - 1) = terminal_velocity;
            else
                acc_current = gravity;
            end
        end

        % Update altitude, pressure, velocity
        data.altitude(i) = data.altitude(i - 1) + data.velocity(i - 1)*dt + 0.5*acc_current*(dt^2);
        data.pressure(i) = altitudeToPressure(data.altitude(i), temp_lapse_rate, temperature, ...
                                              h_b, gravity, molar_mass, gas_constant, static_pressure);
        data.velocity(i) = data.velocity(i - 1) + acc_current*dt;

        prev_t = data.time(i);

        % Stop if near ground level after apogee
        if data.altitude(i) < 1.0 && data.time(i) > time_apogee
            data.num_points = i;
            break;
        end
    end

    % Find apogee
    [max_alt, max_idx] = max(data.altitude);
    data.apogee = max_alt;
    data.time_at_apogee = data.time(max_idx);
end

File: lib/ROCKET/ROCKET_SIM.cpp
#include "ROCKET_SIM.h"

ROCKET_SIM::ROCKET_SIM(float burnout_time, float apogee, float terminal_velocity)
{
    this->burnout_time = burnout_time;
    this->apogee = apogee;
    this->terminal_velocity = terminal_velocity;
}

sim_data &ROCKET_SIM::runSimulation()
{
    // DBG("Initialised constants: " + String(this->burnout_time) + ", " + String(this->apogee) + ", " + String(this->terminal_velocity));

    float acceleration = compute_a_b(this->gravity, this->burnout_time, this->apogee);

    // DBG("Acceleration: " + String(acceleration));

    float time_apogee = abs((this->burnout_time * (acceleration - this->gravity)) / (-this->gravity));

    // DBG("Time to apogee: " + String(time_apogee));

    // time and distance from apogee to terminal velocity
    float time_a_t = abs(this->terminal_velocity / this->gravity);
    float dist_t = this->apogee + ((pow(this->terminal_velocity, 2)) / (2 * this->gravity));

    // time from terminal to landing
    float time_t_l = abs(dist_t / terminal_velocity);

    float time_total = time_apogee + time_a_t + time_t_l;

    // DBG("Total time: " + String(time_total));

    if (time_total <= 0)
    {
        return this->data;
    }

    for (int i = 0; i < resolution; i++)
    {
        this->data.altitude[i] = 0.0f;
        this->data.velocity[i] = 0.0f;
        this->data.pressure[i] = 0.0f;
    }

    float prev_t = 0.0;
    float dt = 0;

    // Previous graph point
    int prev_x = 0;
    int prev_y = 0;

    for (int x = 1; x < resolution; x++)
    {
        this->data.time[x] = mapFloat(x, 1.0f, float(resolution - 1), 0.0f, time_total);
        dt = this->data.time[x] - prev_t;

        if (this->data.time[x] >= this->burnout_time)
        {
            if (this->data.velocity[x - 1] <= this->terminal_velocity)
            {
                acceleration = 0;
                this->data.velocity[x - 1] = this->terminal_velocity;
            }
            else
            {
                acceleration = this->gravity;
            }
        }

        // Update velocity and altitude using kinematic equations
        this->data.altitude[x] = this->data.altitude[x - 1] + (this->data.velocity[x - 1] * dt) + (0.5 * acceleration * pow(dt, 2));
        this->data.pressure[x] = altitudeToPressure(this->data.altitude[x]);
        this->data.velocity[x] = this->data.velocity[x - 1] + acceleration * dt;

        // Update previous points
        prev_x = x;
        prev_y = this->data.altitude[x];
        prev_t = this->data.time[x];

        if (this->data.altitude[x] < 1.0 && this->data.time[x] > time_apogee)
        {
            this->data.num_points = x;
            break;
        }
    }

    // Find the maximum altitude and its index
    float *max_ptr = std::max_element(this->data.altitude, this->data.altitude + this->data.num_points);
    int max_index = static_cast<int>(max_ptr - this->data.altitude);

    // DBG("Max index: " + String(max_index));

    // Retrieve the maximum altitude and corresponding time
    this->data.apogee = this->data.altitude[max_index];

    // DBG("Apogee: " + String(this->data.apogee) + "m");

    this->data.time_at_apogee = this->data.time[max_index];

    return this->data;
}

float ROCKET_SIM::compute_a_b(double g, double t_b, double S_a)
{
    // Calculate -g * t_b^2
    double term1 = -g * t_b * t_b;

    // Calculate the square root argument: (term1)^2 - 4 * t_b^2 * 2gS_a
    double sqrt_argument = (term1 * term1) - (4.0 * t_b * t_b * 2.0 * g * S_a);

    // Check if the sqrt argument is non-negative
    if (sqrt_argument < 0.0)
    {
        // Serial.println("Error: Negative argument inside sqrt. Returning NaN.");
        return NAN; // Not a Number to indicate an error
    }

    // Compute the square root
    double sqrt_value = sqrt(sqrt_argument);

    // Compute the numerator: g * t_b^2 + sqrt(g^2 * t_b^4 - 8gS_a * t_b^2)
    double numerator = g * t_b * t_b + sqrt_value;

    // Compute the denominator: 2 * t_b^2
    double denominator = 2.0 * t_b * t_b;

    // Compute y
    double y = numerator / denominator;

    return y;
}

float ROCKET_SIM::mapFloat(float x, float in_min, float in_max, float out_min, float out_max)
{
    return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
}

float ROCKET_SIM::altitudeToPressure(float h)
{
    // Calculate the base of the exponent
    float base = 1 + ((temp_lapse_rate / temperature) * (h - h_b));

    // Calculate the exponent
    float exponent = (gravity * molar_mass) / (gas_constant * temp_lapse_rate);

    // Calculate final pressure using the formula
    float P = static_pressure * pow(base, exponent);

    return P;
}

float ROCKET_SIM::pressureToAltitude(float P)
{
    // Calculate the exponent for (P / static_pressure) based on rearranged formula
    float exponent = (-gas_constant * temp_lapse_rate) / ((-gravity) * molar_mass);

    // Calculate the term inside the parentheses
    float term = pow(P / static_pressure, exponent) - 1.0f;

    // Calculate altitude based on the rearranged formula
    float h = h_b + (temperature / temp_lapse_rate) * term;

    return h;
}